### PR TITLE
Use docket_number components to lookup and match dockets

### PR DESCRIPTION
**Pull request by albertisfu**
As we agreed, in this PR, I've introduced the use of `docket_number` components in the `find_docket_object` method to improve docket matching.

Here are some thoughts on the approach:

- The `docket_number` components are only used:
    - To refine a match if more than one docket is matched by the current lookups.
    - To confirm the match of a single docket when the `pacer_case_id` is not available, in addition to the `docket_number`.
    - The reason for this decision is that using these components in the main lookups doesn't seem appropriate, as the current dockets in the DB don't contain these new values, causing the lookup to fail. Even when all dockets in the DB have these components, there remains a question: Is it possible to have a `docket_number` like `3:21-cv-00071-B` (`"federal_dn_judge_initials_assigned": "B"`) in one report, but in another report, the `docket_number` `3:21-cv-00071` lacks this data (`"federal_dn_judge_initials_assigned": ""`), leaving the value empty? If so, this could be another reason to avoid using these values in the main lookup.
    - For now, it seems better to use the components if the main lookups (`pacer_case_id + docket_number`) match more than two dockets. Then we can use the DN components to refine the results and select the correct one. Similarly, if a single docket without a `pacer_case_id` is matched, we can use the components to confirm that the `docket_number` is correct.
    - Considering that not all DN components are always available, the refining and confirming queries are performed in a way that only considers DN components with values, avoiding queries with empty values.
    - In the scenario where more than one docket is matched, and after refining by DN components, it's still not possible to find a single match, the current logic applies, selecting the oldest docket as the winner.
- Only three `docket_number` components are considered for these additional lookups: `federal_defendant_number`, `federal_dn_judge_initials_assigned`, and `federal_dn_judge_initials_referred`.
    - The reason is that `federal_dn_case_type` and `federal_dn_office_code` are already included within the `docket_number`, so it would be redundant to use them again in a lookup.
- Finally, when adding the tests, I realized how helpful these additional lookups can actually be, considering we're not parsing these components from non-PACER sources. In district or bankruptcy PACER cases, where the `pacer_case_id` can't be `None` and there can't be more than one case in the same court with the same `pacer_case_id`, the scenarios where the new refining and confirming methods by DN components can be useful are as follows:
    - Duplicated cases due to the lack of `pacer_case_id`. In this case, we might consider duplicated cases both without a `pacer_case_id` (or only one of them), such as one added by a non-PACER source like a docket added by an Opinion. The `docket_number` components can be useful to match the correct docket if both contain the same `docket_number`. However, this could only be helpful if we'd parse `docket_number` components from non-PACER sources as well.
    - A single case matched by a non-PACER source where the `pacer_case_id` is not available. Again, using the DN components can help match the correct docket. But without DN components in non-PACER cases, this logic would never be used.
    - A third hypothetical scenario where this new logic could be useful is if two dockets have the same `pacer_case_id` but different `docket_numbers`. However, I believe this scenario is not possible since `pacer_case_ids` are unique per court.
    - For appellate cases, this logic can also be useful since the `pacer_case_id` can be `None` on appellate dockets. However, we don't have `docket_number` components for appellate cases.

To summarize:

- I can't think of a scenario when merging actual PACER data where using the DN components wouldn't help.
- DN components can be useful for confirming the correct docket when merging dockets from non-PACER sources, provided DN components are available in their `docket_numbers`. But we would need to update the parsers in those sources as well.
- Given that the `pacer_case_id` is mandatory and unique for district and bankruptcy PACER dockets, the new DN components will be added to the correct docket as soon as we receive the new data.

 Let me know what do you think.

### CODE REVIEW

**Review by mlissner — APPROVED** [review]
Yeah, um, this is great. You made all the right assumptions and your code is tidy and clean. No comments here. Merging, thank you!

**Comment by mlissner** [comment]
Automerge is set!
